httpconnect/httpconnect.go

```diff
@@ -147,28 +151,37 @@ func (f *HTTPConnectHandler) intercept(w http.ResponseWriter, req *http.Request)
 			}
 		}
 	}
-	var closeOnce sync.Once
-	go func() {
+
+	readErrCh := make(chan error)
+	op.Go(func() {
 		buf := buffers.Get()
 		defer buffers.Put(buf)
-		if _, err := io.CopyBuffer(connOut, clientConn, buf); err != nil {
-			log.Debug(err)
+		_, readErr := io.CopyBuffer(connOut, clientConn, buf)
+		if readErr != nil {
+			log.Debug(readErr)
 		}
-		closeOnce.Do(closeConns)
-
-	}()
+		readErrCh <- readErr
+	})
 	buf := buffers.Get()
 	defer buffers.Put(buf)
-	if _, err := io.CopyBuffer(clientConn, connOut, buf); err != nil {
-		log.Debug(err)
```

Comment:
The below change of logic was necessary to properly determine success/failure of the request, and it has the added benefit of getting rid of the dreaded "use of closed network connection" error.  The old logic used to close the connection when either reading or writing was finished, even if the other wasn't yet.